File: src/common/JsonObject.cs
using LanguageExt;
using System;
using System.Text.Json.Nodes;

namespace common;

public static class JsonObjectModule
{
    public static JsonResult<JsonNode> GetProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject switch
        {
            null => JsonResult.Fail<JsonNode>("JSON object is null."),
            _ => jsonObject.TryGetPropertyValue(propertyName, out var jsonNode)
                    ? jsonNode switch
                    {
                        null => JsonResult.Fail<JsonNode>($"Property '{propertyName}' is null."),
                        _ => JsonResult.Succeed(jsonNode)
                    }
                    : JsonResult.Fail<JsonNode>($"JSON object does not have a property named '{propertyName}'.")
        };

    public static Option<JsonNode> GetOptionalProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName)
                  .Match(Option<JsonNode>.Some, _ => Option<JsonNode>.None);

    public static JsonResult<T> GetProperty<T>(this JsonObject? jsonObject, string propertyName, Func<JsonNode, JsonResult<T>> selector) =>
        jsonObject.GetProperty(propertyName)
                  .Bind(selector)
                  .AddPropertyNameToErrorMessage(propertyName);

    public static JsonResult<JsonObject> GetJsonObjectProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName, jsonNode => jsonNode.AsJsonObject());

    private static JsonResult<T> AddPropertyNameToErrorMessage<T>(this JsonResult<T> result, string propertyName)
    {
        return result.ReplaceError(replaceError);

        JsonError replaceError(JsonError error) =>
            JsonError.From($"Property '{propertyName}' is invalid. {error.Message}");
    }

    public static JsonResult<JsonArray> GetJsonArrayProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName, jsonNode => jsonNode.AsJsonArray());

    public static JsonResult<JsonValue> GetJsonValueProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName, jsonNode => jsonNode.AsJsonValue());

    public static JsonResult<string> GetStringProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName, jsonNode => jsonNode.AsJsonValue()
                                                                 .Bind(jsonValue => jsonValue.AsString()));

    public static JsonResult<int> GetIntProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName, jsonNode => jsonNode.AsJsonValue()
                                                                 .Bind(jsonValue => jsonValue.AsInt()));

    public static JsonResult<bool> GetBoolProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName, jsonNode => jsonNode.AsJsonValue()
                                                                 .Bind(jsonValue => jsonValue.AsBool()));

    public static JsonResult<Guid> GetGuidProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName, jsonNode => jsonNode.AsJsonValue()
                                                                 .Bind(jsonValue => jsonValue.AsGuid()));

    public static JsonResult<Uri> GetAbsoluteUriProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName, jsonNode => jsonNode.AsJsonValue()
                                                                 .Bind(jsonValue => jsonValue.AsAbsoluteUri()));

    public static JsonObject SetProperty(this JsonObject jsonObject, string propertyName, JsonNode? propertyValue)
    {
        jsonObject[propertyName] = propertyValue;
        return jsonObject;
    }
}

File: src/fsharp/JsonObject.fs
﻿[<RequireQualifiedAccess>]
module common.JsonObject

open FSharpPlus
open System.Text.Json.Nodes

let getProperty propertyName (jsonObject: JsonObject | null) =
    match jsonObject with
    | null -> JsonResult.failWithMessage "JSON object is null."
    | jsonObject ->
        match jsonObject.TryGetPropertyValue(propertyName) with
        | true, property ->
            match property with
            | null -> JsonResult.failWithMessage $"Property '{propertyName}' is null."
            | property -> JsonResult.succeed property
        | _ -> JsonResult.failWithMessage $"Property '{propertyName}' is missing."

let getOptionalProperty propertyName jsonObject =
    getProperty propertyName jsonObject
    |> map Option.Some
    |> JsonResult.defaultWith (fun _ -> Option.None)

let private addPropertyNameToErrorMessage propertyName jsonResult =
    let replaceError jsonError =
        let originalErrorMessage = JsonError.getMessage jsonError

        let newErrorMessage =
            $"Property '{propertyName}' is invalid. {originalErrorMessage}"

        JsonError.fromString newErrorMessage

    JsonResult.replaceErrorWith replaceError jsonResult

let private getPropertyFromResult getPropertyResult propertyName jsonObject =
    getProperty propertyName jsonObject
    |> bind getPropertyResult
    |> addPropertyNameToErrorMessage propertyName

let getJsonObjectProperty propertyName jsonObject =
    getPropertyFromResult JsonNode.asJsonObject propertyName jsonObject

let getJsonArrayProperty propertyName jsonObject =
    getPropertyFromResult JsonNode.asJsonArray propertyName jsonObject

let getJsonValueProperty propertyName jsonObject =
    getPropertyFromResult JsonNode.asJsonValue propertyName jsonObject

let getStringProperty propertyName jsonObject =
    let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asString
    getPropertyFromResult getPropertyResult propertyName jsonObject

let getAbsoluteUriProperty propertyName jsonObject =
    let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asAbsoluteUri
    getPropertyFromResult getPropertyResult propertyName jsonObject

let getGuidProperty propertyName jsonObject =
    let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asGuid
    getPropertyFromResult getPropertyResult propertyName jsonObject

let getBoolProperty propertyName jsonObject =
    let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asBool
    getPropertyFromResult getPropertyResult propertyName jsonObject

let getIntProperty propertyName jsonObject =
    let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asInt
    getPropertyFromResult getPropertyResult propertyName jsonObject

let getDateTimeOffsetProperty propertyName jsonObject =
    let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asDateTimeOffset
    getPropertyFromResult getPropertyResult propertyName jsonObject

let setProperty (propertyName: string) (propertyValue: JsonNode | null) (jsonObject: JsonObject) =
    jsonObject[propertyName] <- propertyValue
    jsonObject

let removeProperty (propertyName: string) (jsonObject: JsonObject) =
    jsonObject.Remove(propertyName) |> ignore
    jsonObject
